src/Stratis.Bitcoin/BlockPulling2/BlockPullerBehavior.cs

```diff
@@ -0,0 +1,245 @@
+            this.averageSizeBytes = new AverageCalculator(samplesCount);
+            this.averageDelaySeconds = new AverageCalculator(samplesCount);
+            this.SpeedBytesPerSecond = 0;
+
+            this.blockPuller = blockPuller;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+            this.loggerFactory = loggerFactory;
+        }
+
+        /// <inheritdoc/>
+        public void AddSample(long blockSizeBytes, double delaySeconds)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:{3})", nameof(blockSizeBytes), blockSizeBytes, nameof(delaySeconds), delaySeconds);
+
+            this.averageSizeBytes.AddSample(blockSizeBytes);
+            this.averageDelaySeconds.AddSample(delaySeconds);
+            
+            this.SpeedBytesPerSecond = (int)(this.averageSizeBytes.Average / this.averageDelaySeconds.Average);
+            
+            this.logger.LogTrace("(-):{0}={1}", nameof(this.SpeedBytesPerSecond), this.SpeedBytesPerSecond);
+        }
+
+        /// <inheritdoc/>
+        public void Penalize(double delaySeconds, int notDeliveredBlocksCount)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:{3})", nameof(delaySeconds), delaySeconds, nameof(notDeliveredBlocksCount), notDeliveredBlocksCount);
+
+            int maxSamplesToPenalize = (int)(this.averageDelaySeconds.GetMaxSamples() * MaxSamplesPercentageToPenalize);
+            int penalizeTimes = notDeliveredBlocksCount < maxSamplesToPenalize ? notDeliveredBlocksCount : maxSamplesToPenalize;
+            
+            this.logger.LogDebug("Peer will be penalized {0} times.", penalizeTimes);
+
+            for (int i = 0; i < penalizeTimes; i++)
+                this.AddSample(0, delaySeconds);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void OnIbdStateChanged(bool isIbd)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(isIbd), isIbd);
+
+            // Recalculates the max samples count that can be used for quality score calculation.
+            int samplesCount = isIbd ? IbdSamplesCount : NormalSamplesCount;
+            this.averageSizeBytes.SetMaxSamples(samplesCount);
+            this.averageDelaySeconds.SetMaxSamples(samplesCount);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void RecalculateQualityScore(int bestSpeedBytesPerSecond)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(bestSpeedBytesPerSecond), bestSpeedBytesPerSecond);
+
+            this.QualityScore = (double)this.SpeedBytesPerSecond / bestSpeedBytesPerSecond;
+
+            if (this.QualityScore < MinQualityScore)
+                this.QualityScore = MinQualityScore;
+
+            if (this.QualityScore > MaxQualityScore)
+                this.QualityScore = MaxQualityScore;
+
+            this.logger.LogTrace("Quality score was set to {0}.", this.QualityScore);
```

Comment:
Do you really need two exit logs?

Reply:
One is the normal log and another one is exit log.